### ISSUE TITLE
[wxMSW / CMake] building in release mode with all samples enabled - fails (sample: memcheck)

### ISSUE THREAD

**Issue by eranif** [wxWidgets/wxWidgets#23106]
I am using `cmake` & `clang` for building wxWidgets.
the build fails when building in `release` mode with this error (coming from file `samples/memcheck.cpp`):

```c++
#ifndef __WXDEBUG__
#error This program must be compiled in debug mode.
#endif
```

this small modification in `samples/CMakeLists.txt` fixes the problem:

```diff
diff --git a/build/cmake/samples/CMakeLists.txt b/build/cmake/samples/CMakeLists.txt
index a016e010a0..a7a8d28e45 100644
--- a/build/cmake/samples/CMakeLists.txt
+++ b/build/cmake/samples/CMakeLists.txt
@@ -89,7 +89,7 @@ wx_add_sample(layout layout.cpp layout.h)
 wx_add_sample(listctrl listtest.cpp listtest.h RES listtest.rc DEPENDS wxUSE_LISTCTRL)
 wx_add_sample(mdi mdi.cpp mdi.h RES mdi.rc DEPENDS wxUSE_MDI wxUSE_DOC_VIEW_ARCHITECTURE wxUSE_MDI_ARCHITECTURE)
 wx_add_sample(mediaplayer LIBRARIES wxmedia DEPENDS wxUSE_MEDIACTRL)
-wx_add_sample(memcheck)
+wx_add_sample(memcheck DEPENDS __WXDEBUG__)
 wx_add_sample(menu DEPENDS wxUSE_MENUS wxUSE_MENUBAR)
 wx_add_sample(minimal IMPORTANT)
 wx_add_sample(notebook notebook.cpp notebook.h LIBRARIES wxaui DEPENDS wxUSE_NOTEBOOK)
```

[cmake.patch](https://github.com/wxWidgets/wxWidgets/files/10329320/cmake.patch)

OS: Windows 10
Compiler / environment: MSYS2/Clang 15.0.2

This is how I build wxWidgets:

```bash
mkdir build-release
cd $_
cmake .. -G"MinGW Makefiles"            \
    -DCMAKE_BUILD_TYPE=Release          \
    -DwxBUILD_DEBUG_LEVEL=0             \
    -DwxBUILD_MONOLITHIC=1              \
    -DwxBUILD_SAMPLES=ALL               \
    -DwxUSE_STL=1                       \
    -DCMAKE_INSTALL_PREFIX=$HOME/root
mingw32-make -j$(nproc) install
```

**Comment by vadz** [comment]
I think this is the right thing to do and should be done in 3.2, but I'm also tempted to remove this sample (and all the related functionality) entirely, I can't believe anybody is still using it in 2022^H3 when there are so many much better tools around.

But for now we should commit this, of course, if @MaartenBent doesn't have any objections.

BTW, Eran, please don't hesitate to submit PRs with your changes, I'm _almost_ sure that this shouldn't break anything but it would still be better to have it tested in the CI builds.

**Comment by MaartenBent** [comment]
I have to disagree, this is not the right thing to do. `__WXDEBUG__` is only defined in the code (`debug.h`), not in the build system. So now the `memcheck` sample is never created.

Also, `__WXDEBUG__` does not depend on Debug or Release build, but on `wxDEBUG_LEVEL`. Which can be set to any value for either configuration.

**Comment by vadz** [comment]
Oh, I thought it was predefined in CMake...

Do you see any simple solution to the problem for 3.2? For master I can/will just remove it as already mentioned.

**Comment by MaartenBent** [comment]
It depends on `wxBUILD_DEBUG_LEVEL`, so this should work:
```cmake
if(wxBUILD_DEBUG_LEVEL STREQUAL "Default" OR wxBUILD_DEBUG_LEVEL GREATER 0)
    wx_add_sample(memcheck)
endif()
```

**Comment by vadz** [comment]
I've finally created a PR with the suggestion above for 3.2. And d74c9e51da (Remove memcheck sample, 2023-01-05) has already removed it from master, so this should be fixed once #23133 is merged.

**Comment by vadz** [comment]
Merged now, so closing.
